virtcontainers/factory/template/template.go

```diff
@@ -100,35 +104,44 @@ func (t *template) createTemplateVM(ctx context.Context) error {
 	config.HypervisorConfig.BootFromTemplate = false
 	config.HypervisorConfig.MemoryPath = t.statePath + "/memory"
 	config.HypervisorConfig.DevicesStatePath = t.statePath + "/state"
+	// template vm uses builtin proxy
+	config.ProxyType = templateProxyType
 
 	vm, err := vc.NewVM(ctx, config)
 	if err != nil {
 		return err
 	}
 	defer vm.Stop()
 
-	err = vm.Pause()
-	if err != nil {
+	if err = vm.Disconnect(); err != nil {
 		return err
 	}
 
-	err = vm.Save()
-	if err != nil {
+	// Sleep a bit to let the agent grpc server clean up
+	time.Sleep(templateWaitForAgent)
+
+	if err = vm.Pause(); err != nil {
+		return err
+	}
+
+	if err = vm.Save(); err != nil {
 		return err
 	}
 
 	// qemu QMP does not wait for migration to finish...
-	time.Sleep(1 * time.Second)
+	time.Sleep(templateWaitForMigration)
```

Comment:
Why a sleep here? Any way we could provide some more robust way of actively waiting?

Reply:
Agree. Also the timeout values seem a bit arbitrary. Is there a possibility that these timeouts may not be sufficient on on slower machines?

Reply:
Could we check the status of the template files being created by `vm.Save()` as a way to validate the migration is done?

Reply:
No, we do not know the actual size of the file either. We might be able to monitor the file close event but it's very hacky and as you said in the other comment, it is on the non-regular path and it does not worth so much effort to optimize.

Reply:
@laijs just mentioned to me that we can use qmp `query-migrate` do check the migration status. Then `templateWaitForMigration` can be dropped. I'll send a new PR to remove it after this one is merged. Does it sound good to you @WeiZhang555 ?

w.r.t. `templateWaitForAgent`, it just affects new normal VM creation speed if your machine is too slow for the agent to come around.

Reply:
`query-migrate` sounds quite good.
Then can you add a "// TODO: " as a reminder? In case everyone of us forgot this, at least in some days someone can find this `TODO` :laughing:

Reply:
@WeiZhang555 here is our #728 ;-)

Reply:
@bergwolf Oh, we already got one! Nice!